demos/ota/ota_demo_core_http/ota_demo_core_http.c

```diff
@@ -0,0 +1,2486 @@
+    HTTPResponse_t xResponse;
+    /* Represents header data that will be sent in an HTTP request. */
+    HTTPRequestHeaders_t xRequestHeaders;
+
+    /* Return value of all methods from the HTTP Client library API. */
+    HTTPStatus_t xHttpStatus = HTTPSuccess;
+
+    /* Reconnection required flag. */
+    bool xReconnectRequired = false;
+
+    /* Initialize all HTTP Client library API structs to 0. */
+    ( void ) memset( &xRequestInfo, 0, sizeof( xRequestInfo ) );
+    ( void ) memset( &xResponse, 0, sizeof( xResponse ) );
+    ( void ) memset( &xRequestHeaders, 0, sizeof( xRequestHeaders ) );
+
+    /* Initialize the request object. */
+    xRequestInfo.pHost = pcServerHost;
+    xRequestInfo.hostLen = xServerHostLength;
+    xRequestInfo.pMethod = HTTP_METHOD_GET;
+    xRequestInfo.methodLen = sizeof( HTTP_METHOD_GET ) - 1;
+    xRequestInfo.pPath = pcPath;
+    xRequestInfo.pathLen = strlen( pcPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. */
+    xRequestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    xRequestHeaders.pBuffer = pucHttpUserBuffer;
+    xRequestHeaders.bufferLen = HTTP_USER_BUFFER_LENGTH;
+
+    xHttpStatus = HTTPClient_InitializeRequestHeaders( &xRequestHeaders,
+                                                       &xRequestInfo );
+
+    HTTPClient_AddRangeHeader( &xRequestHeaders, ulRangeStart, ulRangeEnd );
+
+    if( xHttpStatus == HTTPSuccess )
+    {
+        /* Initialize the response object. The same buffer used for storing
+         * request headers is reused here. */
+        xResponse.pBuffer = pucHttpUserBuffer;
+        xResponse.bufferLen = HTTP_USER_BUFFER_LENGTH;
+
+        /* Send the request and receive the response. */
+        xHttpStatus = HTTPClient_Send( &xTransportInterfaceHttp,
+                                       &xRequestHeaders,
+                                       NULL,
+                                       0,
+                                       &xResponse,
+                                       0 );
+    }
+    else
+    {
+        LogError( ( "Failed to initialize HTTP request headers: Error=%s.",
+                    HTTPClient_strerror( xHttpStatus ) ) );
+    }
+
+    if( xHttpStatus != HTTPSuccess )
+    {
+        if( ( xHttpStatus == HTTPNoResponse ) || ( xHttpStatus == HTTPNetworkError ) )
+        {
+            xReconnectRequired = true;
+        }
+        else
+        {
+            LogError( ( "HTTPClient_Send failed: Error=%s.",
+                        HTTPClient_strerror( xHttpStatus ) ) );
+
+            xReturnStatus = OtaHttpRequestFailed;
+        }
+    }
+    else
+    {
+        /* Check if reconnection required. */
+        if( xResponse.respFlags & HTTP_RESPONSE_CONNECTION_CLOSE_FLAG )
+        {
+            xReconnectRequired = true;
+        }
```

Comment:
The HTTP_RESPONSE_CONNECTION_CLOSE_FLAG could be received after successfully parsing a response right? Do we always want to set this to true then?

Reply:
`xResponse.respFlags` denotes the response of the server and is accurate for determining if we need to re-establish the connection

Reply:
With the current implementation, wouldn't this make the device attempt to reconnect again even after successfully receiving the end of the response, even when we don't want to receive any more data?

It seems like it would reach the end of the communication, and then try to reconnect until it reaches the max attempts. Right now I think this is just one attempt which is probably why we don't have any issues. If we update this, we could see a significant delay being introduced here.

Reply:
Ah, I see your point. It would take some extra time to reconnect, if it disconnects after downloading the entire file. However, once the file is completely downloaded we call OTA_ActivateNewImage() which would reset the device and the delay can be ignored

Reply:
That makes sense, this is not an immediate concern then. I think it would be best to avoid this unnecessary connection if possible though. If there is a way that we can detect if this is the end of a connection and then avoid reconnecting, that would be great. There are maybe some scenarios where we wouldn't reset. For example if the device has an issue with one of the blocks or we receive a new job. Effectively just some scenario where we reject the current job.

Reply:
Yes, this is an edge case and can be handled at a later stage after more testing.